storage: add test to verify replicasByKey behavior

rjnn opened the pull request:
When a store accepts a snapshot for a range that split or otherwise
changed its descriptor, it should update its internal state. Fixes
#8946.

WIP as the test is scaffolded but nonfunctional. I would appreciate 
any feedback on how to make that test work.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8954)

<!-- Reviewable:end -->

tamird commented:
Be sure to include `#` when mentioning the issue number or else GitHub won't close issues for you.

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/store.go, line 1785 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KQRu6bINum0FRDpIcaA:-KQRu6bINum0FRDpIcaB:b-xvx34y) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00f6b60b9ccade3b535be28d815b2f83aa5ef90/storage/store.go#L1785)):*
> ```Go
> 				s.LookupReplica(rngDesc.StartKey, nil))
> 			return nil
> 		}
> ```

Why do you have to lookup the replica here? Isn't using `rng` (the parameter to this method) sufficient? But then that would obviate the need for any of this code. So how can we get here such that `s.mu.replicas[rangeID] != rng`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/client_split_test.go, line 1078 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KQSJp6Wg_mvMWWGVp5G:-KQSJp6Xwkv3MwpADevt:buotxrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00f6b60b9ccade3b535be28d815b2f83aa5ef90/storage/client_split_test.go#L1078)):*
> ```Go
> func TestSplitSnapshotSplitWins_RangeDescUpdate(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> 	mtc, leftKey, rightKey := setupSplitSnapshotRace(t)
> ```

I think you don't need the whole split snapshot race here. With three nodes, you can bring the third one down, have the other two apply some commands (including a split), then force a log truncation (so the third node will have to get a snapshot when it comes back). The tricky part is what to test for after the snapshot has been applied - that stuff isn't exposed to the `client` package. Semi-related idea - we could expose something on the store that lets it run internal consistency checks (one that would catch the memory corruption that this test would create prior to a fix) and then run that proactively in this test (and in production). This could complement the replica consistency checker, which is mostly about on-disk state, and various assertions we have on *Replica, but not on *Store.

---

*[storage/store.go, line 1767 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KQSJ3fYh2rcPQDyBtj3:-KQSJ3fZ1bb0quugsxZv:ba9wydx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00f6b60b9ccade3b535be28d815b2f83aa5ef90/storage/store.go#L1767)):*
> ```Go
> }
> 
> // processRangeDescriptorUpdateLocked requires that the store lock is held.
> ```

While you're here, could you add more commentary on when this should be called?

---

*[storage/store.go, line 1785 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KQRu6bINum0FRDpIcaA:-KQSJGeJ8yp9ST7G5UkR:bgzlw55) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00f6b60b9ccade3b535be28d815b2f83aa5ef90/storage/store.go#L1785)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Why do you have to lookup the replica here? Isn't using `rng` (the parameter to this method) sufficient? But then that would obviate the need for any of this code. So how can we get here such that `s.mu.replicas[rangeID] != rng`?
</details>
I think the intention is  to get the old descriptor to be able remove from the replicasByKey btree, though that won't work because this returns the exact same replica object (right?)

I just saw that https://github.com/cockroachdb/cockroach/issues/8946 mentions this problem, so you're already aware of it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954)*
<!-- Sent from Reviewable.io -->

Commit:
storage: add test to verify replicasByKey behavior

store.replicasByKey is keyed on replica StartKey and EndKey, directly accessing
those values during BTree traversal. Add a test that verifies that when those
keys are modified in the underlying replicas, this is visible on operations
that rely upon the BTree. Closes #8946.

arjunravinarayan commented:
PTAL @petermattis, changed to just a test that verifies the behavior we expect from replicasByKey, with no changes to the store code. Let me know if I should close this PR and open a new one.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-:-KQalJ6jf_bTGjy_wmaU:b1o9dds)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 2 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[storage/store_test.go, line 375 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KQbMAI3M0NNIGIasGnb:-KQbMAI40RrG8FwKyd5B:bgfdr29) ([raw file](https://github.com/cockroachdb/cockroach/blob/20016cba2864a2247e1092a6b201b293d58c9cb4/storage/store_test.go#L375)):*
> ```Go
> 	defer stopper.Stop()
> 
> 	// Remove the main replica so that we can test on that key space.
> ```

seems like you're removing this replica just to replace it with another replica that you mutate. perhaps just mutate this one?

I also think you can get rid of the list of 4 replicas below, only 2 of which seem to matter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/store_test.go, line 400 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KQe3Fem_XGw8zTxc_RI:-KQe3Fem_XGw8zTxc_RJ:bxerhte) ([raw file](https://github.com/cockroachdb/cockroach/blob/20016cba2864a2247e1092a6b201b293d58c9cb4/storage/store_test.go#L400)):*
> ```Go
> 		desc.replica = createReplica(store, roachpb.RangeID(desc.id), desc.start, desc.end)
> 		if err := store.AddReplicaTest(desc.replica); (desc.expectedErrorOnAdd && err == nil) || (!desc.expectedErrorOnAdd && err != nil) {
> 			t.Fatalf("expected %t, but received %v", desc.expectedErrorOnAdd, err)
> ```

This will give weird error messages like "expected true but received nil". I'd handle the expected-error and expected-no-error branches separately with their own error messages. Also be sure to include the loop index for any error message in a loop. 

---

*[storage/store_test.go, line 408 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KQe4B36cMIwKzf-w0__:-KQe4B36cMIwKzf-w0_a:b5nslnj) ([raw file](https://github.com/cockroachdb/cockroach/blob/20016cba2864a2247e1092a6b201b293d58c9cb4/storage/store_test.go#L408)):*
> ```Go
> 	}
> 
> 	// Modify a range in place
> ```

Not all modifications are safe - increasing EndKey may collide with another range, while decreasing it never does (unless it is decreased below its StartKey). We should be testing the patterns that will occur in the real world. Instead of removing reps[3], and expanding reps[1], the test should shrink reps[1] and then do the tests below with ranges [d-e) and [e-f). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 1767 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KQSJ3fYh2rcPQDyBtj3:-KR-9DiznOOnCtgrUALL:b-lcgf7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00f6b60b9ccade3b535be28d815b2f83aa5ef90/storage/store.go#L1767)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> While you're here, could you add more commentary on when this should be called?
</details>
Added some for the non-Locked version just above.

---

*[storage/store_test.go, line 375 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KQbMAI3M0NNIGIasGnb:-KR-Smrmgm0PRG5Mh5YN:b-n24dbv) ([raw file](https://github.com/cockroachdb/cockroach/blob/20016cba2864a2247e1092a6b201b293d58c9cb4/storage/store_test.go#L375)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> seems like you're removing this replica just to replace it with another replica that you mutate. perhaps just mutate this one?
> 
> I also think you can get rid of the list of 4 replicas below, only 2 of which seem to matter.
</details>
Refactored to use fewer replicas (but also to deal with Ben's comments below). PTAL again.

---

*[storage/store_test.go, line 400 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KQe3Fem_XGw8zTxc_RI:-KR-SieWgGkM-ELLLa2V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20016cba2864a2247e1092a6b201b293d58c9cb4/storage/store_test.go#L400)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This will give weird error messages like "expected true but received nil". I'd handle the expected-error and expected-no-error branches separately with their own error messages. Also be sure to include the loop index for any error message in a loop. 
</details>
Done.

---

*[storage/store_test.go, line 408 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KQe4B36cMIwKzf-w0__:-KR-SlHtXT7QMWdFlGPY:buwjwld) ([raw file](https://github.com/cockroachdb/cockroach/blob/20016cba2864a2247e1092a6b201b293d58c9cb4/storage/store_test.go#L408)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Not all modifications are safe - increasing EndKey may collide with another range, while decreasing it never does (unless it is decreased below its StartKey). We should be testing the patterns that will occur in the real world. Instead of removing reps[3], and expanding reps[1], the test should shrink reps[1] and then do the tests below with ranges [d-e) and [e-f). 
</details>
Done. I'm testing shrinkages now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 1764 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KR-VS9RSlr2wR9BBZDC:-KR-VS9RSlr2wR9BBZDD:b-cyf59q) ([raw file](https://github.com/cockroachdb/cockroach/blob/4889dc37705cd19d80ab2dfb36d46c688ed22094/storage/store.go#L1764)):*
> ```Go
> }
> 
> // processRangeDescriptorUpdate should be called whenever a range's descriptor
> ```

a replica's range descriptor

---

*[storage/store_test.go, line 389 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KR-Vl832uBI9k3RjqhI:-KR-Vl832uBI9k3RjqhJ:bq0vxv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4889dc37705cd19d80ab2dfb36d46c688ed22094/storage/store_test.go#L389)):*
> ```Go
> 		id                 int
> 		start, end         roachpb.RKey
> 		expectedErrorOnAdd bool
> ```

replace this boolean with the expected error so that you can use `testutils.IsError` below rather than the error-prone nil check.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions.

---

*[storage/store.go, line 1764 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KR-VS9RSlr2wR9BBZDC:-KR-Y7ndC_SfkBtsiHXE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4889dc37705cd19d80ab2dfb36d46c688ed22094/storage/store.go#L1764)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> a replica's range descriptor
</details>
Done.

---

*[storage/store_test.go, line 389 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KR-Vl832uBI9k3RjqhI:-KR-Z-M8mMmfHrDPdUL7:b-vqszkn) ([raw file](https://github.com/cockroachdb/cockroach/blob/4889dc37705cd19d80ab2dfb36d46c688ed22094/storage/store_test.go#L389)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> replace this boolean with the expected error so that you can use `testutils.IsError` below rather than the error-prone nil check.
</details>
Can't use `testutils.IsError` to check for the non-error case. I'd still need the bool. Added a new field.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/store_test.go, line 389 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KR-Vl832uBI9k3RjqhI:-KR-d78c6pLGZr_nR_MR:b-4qqs17) ([raw file](https://github.com/cockroachdb/cockroach/blob/4889dc37705cd19d80ab2dfb36d46c688ed22094/storage/store_test.go#L389)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> Can't use `testutils.IsError` to check for the non-error case. I'd still need the bool. Added a new field.
</details>
Sure you can - use the empty string to indicate that no error is expected. This pattern pervades our code.

```
err := ...
if desc.expectedErr == "" {
  if err != nil {
    t.Fatalf(...)
  }
} else {
  if !testutils.IsError(desc.expectedErr) {
    t.Fatalf(...)
  }
}
```

In fact, this pattern is so common that we should probably fold it into `IsError` =/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Update the PR title to match the new commit message.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[storage/store_test.go, line 369 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KR1LZgvybTCPBO2K1Kj:-KR1LZgwFs587P9widSL:btw347i) ([raw file](https://github.com/cockroachdb/cockroach/blob/4972a46aab08afafc48f5dcd17c293cd761865ee/storage/store_test.go#L369)):*
> ```Go
> // TestReplicasByKey tests that operations that depend on the
> // store.replicasByKey map function correctly when the underlying replicas'
> // start and end keys are manipulated in place.
> ```

Say that this happens when a snapshot is applied which advances the replica past a split.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-:-KR1LmvBH09rDVTBPKpn:bnnt2wi)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Done, also rebased. 

---

Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[storage/store_test.go, line 389 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KR-Vl832uBI9k3RjqhI:-KR4FFVqAikg_NbswJ7S:bnhltto) ([raw file](https://github.com/cockroachdb/cockroach/blob/4889dc37705cd19d80ab2dfb36d46c688ed22094/storage/store_test.go#L389)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Sure you can - use the empty string to indicate that no error is expected. This pattern pervades our code.
> 
> ```
> err := ...
> if desc.expectedErr == "" {
>   if err != nil {
>     t.Fatalf(...)
>   }
> } else {
>   if !testutils.IsError(desc.expectedErr) {
>     t.Fatalf(...)
>   }
> }
> ```
> 
> In fact, this pattern is so common that we should probably fold it into `IsError` =/
</details>
Done. I agree that it should be folded into `IsError`.

---

*[storage/store_test.go, line 369 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-KR1LZgvybTCPBO2K1Kj:-KR4FDKRoSaKegVBrywh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4972a46aab08afafc48f5dcd17c293cd761865ee/storage/store_test.go#L369)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Say that this happens when a snapshot is applied which advances the replica past a split.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-:-KR4KyWOtyNnHiO9k5vp:bxlzr10)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8954#-:-KR4ccdxeUZq5xE7bkH4:bnfp4nl)*
<!-- Sent from Reviewable.io -->
